### PR TITLE
Additional Organizational Units ACLs

### PR DESCRIPTION
**Pull request by q-roland**
## Description

This pull request adds the display of two additional edges related to Organizational Units ACLs: **GenericWrite** and **Manage Group Policy Links**. In addition, a warning regarding the exploitation of OU ACL inheritance in the case of adminCount=1 protected objects was added in the OU GenericAll edge abuse information.

## Motivation and Context
The motivation behind the pull request is described in the following article: https://www.synacktiv.com/publications/ounedpy-exploiting-hidden-organizational-units-acl-attack-vectors-in-active-directory
This change aims to add two potentially exploitable OU ACLs to BloodHound. Note that two pull requests were created on the SharpHound collector and on the BloodHound.py collector in order to add the collection of these permissions. 

## How Has This Been Tested?

Various domain extracts with the modified SharpHound/BloodHound.py collectors were performed and ingested in the modified BloodHound version. New edges are displayed as expected. Abuse information for the new edges is also reported in the expected format.

## Screenshots (if appropriate):

https://www.synacktiv.com/sites/default/files/inline-images/bh_patched_managegplink_2.png
https://www.synacktiv.com/sites/default/files/inline-images/bh_patched_genericwrite_2.png
https://www.synacktiv.com/sites/default/files/inline-images/bh_patched_genericall_2.png

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [ ] I have added and/or updated tests to cover my changes.
-   [ ] All new and existing tests passed.
-   [ ] My changes include a database migration.

About this part, I do not think that documentation updates are necessary, nor the addition of tests, but feel free to correct me if I am wrong !

I am at your disposal for any further information

### CODE REVIEW

**Comment by JonasBK** [comment]
Hi @q-roland,

Thank you for publishing a great blog post and submitting this awesome pull request!

I have a couple of questions:
1. The "Manage Group Policy Links" permission creates two ACEs - Write access to the gPLink attribute and the gPOptions. The SharpHoundCommon PR looks for just write access to the gPLink attribute. Based on how I read your blog post, it seems like write access to the gPLink of the targeted OU is enough to pull off the attack. Can you confirm that?
2. Do you know if it is possible to pull off the attack by creating an SMB share and running an LDAP server on a compromised existing AD computer, so you don't rely on being able to create a new computer account in AD?

Thanks,
Jonas

**Comment by q-roland** [comment]
Hi @JonasBK,

Thanks a lot for your answer, I'm glad the SpecterOps team considers the pull request as pertinent!

Regarding your questions:

1. Indeed, the "Manage Group Policy Links" permission also creates an ACE for the gPOptions attribute. This attribute takes the form of a bitmask may be used to apply specific configurations when linking a GPO to an OU (disable user configuration, disable machine configuration, loopback processing, override and auditing etc.). The default value for the gPOptions attribute is 0x00000000, meaning that no specific configuration is applied. In most cases, the exploit described in the article will work by only modifying the gPLink attribute, so the answer to your question is "yes, write access to the gPLink of the targeted OU is enough to pull off the attack". In very specific edge cases, an alteration of the gPOptions attribute may be however necessary - typically, if gPOptions is configured so that user configuration is disabled for an OU, and the attacker wishes to target a user object located in this OU (pretty unlikely though). To sum up, the answer to your first question is "Yes, 99% of the time".
2. Indeed, it is totally possible to implement the attack using an already compromised domain machine to run an LDAP and SMB server. The OUned.py tool proposes an implementation from a non-domain joined machine to minimize attack prerequisites, but using a compromised machine can be also considered.

Do not hesitate if there is any other question,
Quentin

**Review by JonasBK — CHANGES_REQUESTED** [review]
Hi @q-roland,

I've reviewed your pull request more thoroughly now, and it's excellent work! I've included a set of comments outlining some requested changes. If you have any questions about the comments, feel free to ask. Also, please let me know if you'd like to address those changes yourself or if you'd prefer us to handle them. Either way, your code will be merged in, and we truly appreciate your contribution! 😊

**Comment by q-roland** [comment]
Hi @JonasBK,

Thanks a lot for the PR review and the proposed changes. All of them are really relevant - some even provide additional elements that I did not think about (such as abusing write permissions on gPLink by linking a modified GPO to an OU, which is a good point!).

The next few days may be a bit busy for me. If it is possible to implement the changes on your side, it would be great! (Otherwise no worries, I can take the time to do it in the middle of next week).

Thanks again for your input,
Quentin

File: packages/cue/bh/ad/ad.cue [hunk 1]
```diff
@@ -899,6 +899,11 @@ WriteAccountRestrictions: types.#Kind & {
 	schema: "active_directory"
 }
 
+ManageGPLink: types.#Kind & {
```
[the remark this answers is not in the document]
Comment (q-roland) [reply packages/cue/bh/ad/ad.cue:902]: WriteGPLink is indeed a more coherent name.

File: packages/javascript/bh-shared-ui/src/components/HelpTexts/GenericAll/LinuxAbuse.tsx [hunk 2]
```diff
@@ -399,6 +399,32 @@ const LinuxAbuse: FC<EdgeInfoProps & { targetId: string; haslaps: boolean }> = (
                         right you want to apply to precisely which kinds of descendent objects. Refer to the Windows
                         Abuse info for this.
                     </Typography>
+
+                    <Typography variant='body1'>Objects with AdminCount=1 property</Typography>
```
[the remark this answers is not in the document]
Comment (q-roland) [reply packages/javascript/bh-shared-ui/src/components/HelpTexts/GenericAll/LinuxAbuse.tsx:403]: I did not consider other situations in which ACL inheritance may be disabled but this is absolutely correct.

File: packages/javascript/bh-shared-ui/src/components/HelpTexts/GenericAll/LinuxAbuse.tsx [hunk 3]
```diff
@@ -399,6 +399,32 @@ const LinuxAbuse: FC<EdgeInfoProps & { targetId: string; haslaps: boolean }> = (
...
+                        It may still be possible to exploit GenericAll permissions on an Organizational Unit through an alternative attack vector. Indeed, with GenericAll 
+                        permissions over an Organizational Unit, you may make modifications to the gPLink attribute of the OU. 
+                        The ability to alter the gPLink attribute of an Organizational Unit may allow an attacker to apply a malicious 
+                        Group Policy Object to all of the OU's child items (including the ones included in nested sub-OUs). This can be exploited 
+                        to make said child items execute arbitrary commands through an immediate scheduled task, thus compromising them. 
+                    </Typography>
+
+                    <Typography variant='body2'>
+                        Successful exploitation will require the possibility to add non-existing DNS records to the domain and to 
+                        create machine accounts.
```
[the remark this answers is not in the document]
Comment (q-roland) [reply packages/javascript/bh-shared-ui/src/components/HelpTexts/GenericAll/LinuxAbuse.tsx:420]: This is a very good observation - linking a modified GPO to a sensitive OU may allow compromising privileged objects from edit rights on any GPO.

**Comment by JonasBK** [comment]
Hi @q-roland,

Glad to hear my suggestions make sense to you! If you're willing to implement the changes, we're happy to wait for you to find the time. No stress! :)

Cheers,
Jonas

**Comment by q-roland** [comment]
Hi @JonasBK,

I hope you are doing well. I implemented the requested changes in my latest commit. This includes the renaming of "ManageGPLink" to "WriteGPLink", as well as the improvement of Help Texts related to GPLink exploitation, based on your suggestions.

The necessary changes were also committed to the pull request on the SharpHoundCommon repository. I performed some additional tests to ensure the modifications did not break anything.

Do not hesitate to get back to me if some of the changes are incomplete / if you have other suggestions regarding the pull request :)

Cheers,
Quentin

**Comment by JonasBK** [comment]
Hey @q-roland,

Wonderful! I will review it and let you know :)

Thanks for all your great work!

Cheers

**Comment by JonasBK** [comment]
Hey @q-roland,

It looks great! 

Something I have not thought about before now - have you checked if the attack is possible with write gPLink permission on the domain head object? I guess it should work as well.

**Comment by q-roland** [comment]
Hi Jonas,

This is again an interesting point. I was indeed able to reproduce the attack this afternoon by modifying the gPLink attribute of the domain itself. I actually added a malicious GPO to the gPLink value, and in addition to the legitimate Default Domain Policy, all domain objects are now attempting to apply the malicious GPO hosted on the attacker machine.

The exploitation of the domain ACL would obviously have an important impact and could directly result in full domain takeover. However:
-  Permissions on the gPLink attribute seems to be appropriately restricted to administrative users. I do not really see a specific case in which this configuration would be changed and result in less-privileged users having write permissions on the domain gPLink, so exploitation scenarios seems rather unlikely (although it is true that we sometimes encounter unexpected and weird permissions in AD environments).
- Manipulations related to the default domain policy are always a little touchy - even though in this case I exploited it with OUned.py, and nothing seemed to break.
- This attack may be pretty noisy and CPU intensive depending on the network configuration of the target network. Indeed, numerous objects may attempt to fetch their GPO on the attacker's machine. A lot of them may not be successful if they can't reach said attacker machine.

This is still a valid and potentially impactful gPLink abuse scenario. Correct me if I'm wrong, but this is probably not implemented in the current pull request and that may require some additional development, since I suppose that compromise paths related to the domain container are distinct from OU compromise paths

**Comment by JonasBK** [comment]
Hey Quentin,

Awesome, thanks for confirming that it works against domains as well.

Those are some good considerations around the attack. I think it makes sense to add the WriteGPLink and GenericWrite edges against domains as the impact is huge if they exist. I will add a note in the abuse info about that the user should be mindful of what and how many users and computers are affected. 

It is possible to limit the scope of affected principals on a GPO: 
image
I am not sure how it works under the hood, but it could potentially be a cool feature to add to OUned.py if it is feasible.

I will make the changes to BloodHound and SharpHoundCommon - you have already helped a lot with this :)

**Review by JonasBK — APPROVED** [review]

**Pull request merged**